sdk/ai/azure-ai-projects/azure/ai/projects/tracing/agents/_ai_agents_instrumentor.py

```diff
@@ -0,0 +1,1492 @@
+
+    This class allows enabling or disabling tracing for AI Agents.
+    and provides functionality to check whether instrumentation is active.
+
+    """
+
+    def __init__(self):
+        if not _tracing_library_available:
+            raise ModuleNotFoundError(
+                "Azure Core Tracing Opentelemetry is not installed. "
+                "Please install it using 'pip install azure-core-tracing-opentelemetry'"
+            )
+        # In the future we could support different versions from the same library
+        # and have a parameter that specifies the version to use.
+        self._impl = _AIAgentsInstrumentorPreview()
+
+    def instrument(self) -> None:
+        """
+        Enable trace instrumentation for AI Agents.
+
+        Raises:
+            RuntimeError: If instrumentation is already enabled.
+
+        """
+        self._impl.instrument()
+
+    def uninstrument(self) -> None:
+        """
+        Remove trace instrumentation for AI Agents.
+
+        Raises:
+            RuntimeError: If instrumentation is not currently enabled.
+
+        This method removes any active instrumentation, stopping the tracing
+        of AI Agents.
+        """
+        self._impl.uninstrument()
+
+    def is_instrumented(self) -> bool:
+        """
+        Check if trace instrumentation for AI Agents is currently enabled.
+
+        :return: True if instrumentation is active, False otherwise.
+        :rtype: bool
+        """
+        return self._impl.is_instrumented()
+
+
+class _AIAgentsInstrumentorPreview:
+    """
+    A class for managing the trace instrumentation of AI Agents.
+
+    This class allows enabling or disabling tracing for AI Agents.
+    and provides functionality to check whether instrumentation is active.
+    """
+
+    def _str_to_bool(self, s):
+        if s is None:
+            return False
+        return str(s).lower() == "true"
+
+    def instrument(self):
+        """
+        Enable trace instrumentation for AI Agents.
+
+        Raises:
+            RuntimeError: If instrumentation is already enabled.
+
+        This method checks the environment variable
+        'AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED' to determine
+        whether to enable content tracing.
+        """
+        if self.is_instrumented():
+            raise RuntimeError("Already instrumented")
+
+        var_value = os.environ.get("AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED")
+        enable_content_tracing = self._str_to_bool(var_value)
+        self._instrument_agents(enable_content_tracing)
+
+    def uninstrument(self):
+        """
+        Disable trace instrumentation for AI Agents.
+
+        Raises:
+            RuntimeError: If instrumentation is not currently enabled.
+
+        This method removes any active instrumentation, stopping the tracing
+        of AI Agents.
+        """
+        if not self.is_instrumented():
+            raise RuntimeError("Not instrumented")
```

Comment:
this can help increase quality of client code by making unnecessary calls more obvious

Reply:
as I mention in the other comment, this will be hard for end-users to control + it's a common practice that you can close the same resource over and over again without raising any errors.

Reply:
could you point to some examples?

Reply:
https://docs.python.org/3/library/io.html#io.IOBase.close

> Flush and close this stream. This method has no effect if the file is already closed. Once the file is closed, any operation on the file (e.g. reading or writing) will raise a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError).

As a convenience, it is allowed to call this method more than once; only the first call, however, will have an effect.

https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

> To help ensure that resources are always cleaned up appropriately, a [Dispose](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable.dispose) method should be idempotent, such that it's callable multiple times without throwing an exception. Furthermore, subsequent invocations of [Dispose](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable.dispose) should do nothing.

https://antonz.org/idempotent-close/

Reply:
https://docs.oracle.com/javase/8/docs/api/index.html?java/lang/AutoCloseable.html

> Note that unlike the [close](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html#close--) method of 
[Closeable](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html), this close method is not required to be idempotent. In other words, calling this close method more than once may have some visible side effect, unlike Closeable.close which is required to have no effect if called more than once. However, implementers of this interface are strongly encouraged to make their close methods idempotent.

Reply:
changed